aesara/tensor/math_opt.py

```diff
@@ -226,6 +227,42 @@ def local_func_inv(fgraph, node):
     return
 
 
+@register_canonicalize
+@register_specialize
+@local_optimizer([Elemwise])
+def local_log_exp(fgraph, node):
+    """
+    Convert ``aet.log(aet.exp(x))`` to ``x`` and ``aet.exp(aet.log(x))`` to
+    ``aet.switch(x >= 0, x, np.nan)``.
+    """
+
+    x = node.inputs[0]
+
+    if not isinstance(node.op, Elemwise):
+        return
+    if not x.owner or not isinstance(x.owner.op, Elemwise):
+        return
+
+    prev_op = x.owner.op.scalar_op
+    node_op = node.op.scalar_op
+
+    if isinstance(prev_op, aes.Exp) and isinstance(node_op, aes.Log):
+        return x.owner.inputs
+
+    if isinstance(prev_op, aes.Log) and isinstance(node_op, aes.Exp):
+        out = node.outputs[0]
+
+        x = x.owner.inputs[0]
+        new_out = switch(ge(x, 0), x, np.nan)
+
+        if new_out.dtype != out.dtype:
```

Comment:
Not sure if this is needed. Does switch ever change the dtype of it's inputs?

Reply:
I don't believe there is a 32bit NaN, so that should always upcast to `float64`.

Reply:
float32 `nan` seems fine:
```python
x = at.fvector('x')
y = at.cast(at.switch(x > 0, x, np.nan), 'float32')
f = aesara.function([x], y)

print( f([-5.0]), f([-5.0]).dtype)
# [nan] float32
```

Reply:
As long as there's a usable 16/32 `nan`, then I'm fine with the cast to the original `dtype`.